### PR TITLE
Fix Service Update on LoadBalancerSourceRanges Field

### PR DESCRIPTION
**Pull request by freehan**
Fixes: https://github.com/kubernetes/kubernetes/issues/33033
Also expands: https://github.com/kubernetes/kubernetes/pull/32748

### CODE REVIEW

**Comment by k8s-oncall** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37720)

**Comment by freehan** [comment]
Applying P2 for catching 1.4.8 release

File: test/e2e/service.go [hunk 1]
```diff
@@ -1103,6 +1103,74 @@ var _ = framework.KubeDescribe("Services", func() {
 			framework.Failf("expected un-ready endpoint for Service %v within %v, stdout: %v", t.name, kubeProxyLagTimeout, stdout)
 		}
 	})
+
+	It("should only allow access from service loadbalancer source ranges [Slow]", func() {
+		// this feature currently supported only on GCE/GKE/AWS
+		framework.SkipUnlessProviderIs("gce", "gke", "aws")
+
+		loadBalancerCreateTimeout := loadBalancerCreateTimeoutDefault
+		largeClusterMinNodesNumber := 100
```
Comment (bprashanth) [inline test/e2e/service.go]: isn't there already a package level const you can use for this?

**Review by bprashanth — COMMENTED** [review]
LGTM random nits

**Comment by freehan** [comment]
I observed a flake after running the test overnight. Will come back

File: pkg/controller/service/servicecontroller.go [hunk 2]
```diff
@@ -430,6 +430,13 @@ func (s *ServiceController) needsUpdate(oldService *v1.Service, newService *v1.S
 			oldService.Spec.Type, newService.Spec.Type)
 		return true
 	}
+
+	if wantsLoadBalancer(newService) && !reflect.DeepEqual(oldService.Spec.LoadBalancerSourceRanges, newService.Spec.LoadBalancerSourceRanges) {
+		s.eventRecorder.Eventf(newService, v1.EventTypeNormal, "LoadBalancerSourceRanges", "%v -> %v",
```
Comment (freehan) [inline pkg/controller/service/servicecontroller.go]: Just confirmed by manual testing, only firewall update happens.

**Comment by freehan** [comment]
The test flake is red herring. Applying LGTM based on previous comment.

**Comment by tonglil** [comment]
Thank you for this! Does this mean mutating `loadBalancerSourceRanges` will _no longer_ show as?
```
spec.loadBalancerSourceRanges: Invalid value: ["1.2.3.4/24","5.6.7.8/24"]: field is immutable
```

**Comment by freehan** [comment]
@tonglil  Yes.

**Comment by tonglil** [comment]
Thanks @freehan. Will this be backported into 1.4.*? Or set for 1.5?

**Comment by freehan** [comment]
@tonglil Will do both. I want to let it sit over the weekend. See if any test flakes happens.

**Comment by tonglil** [comment]
Thanks for the information! 👍

**Comment by sttts** [comment]
These changes cause a P0 flake: https://github.com/kubernetes/kubernetes/issues/38174

**Comment by sttts** [comment]
Looks like the cherry-pick to the 1.4 release branch breaks the branch.

@jessfraz ^^

**Comment by freehan** [comment]
Checking

**Comment by freehan** [comment]
The master is still running 1.4.6. Hence the code was not present on the master. 
```
W1212 03:06:27.522] 2016/12/12 03:06:27 e2e.go:934: Running: kubectl version
I1212 03:06:27.581] Client Version: version.Info{Major:"1", Minor:"4+", GitVersion:"v1.4.8-beta.0", GitCommit:"ba5b2fcee9d938dae74d4474f68d9d0933f3ec04", GitTreeState:"clean", BuildDate:"2016-12-10T05:48:32Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
I1212 03:06:27.643] Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.6", GitCommit:"e569a27d02001e343cb68086bc06d47804f62af6", GitTreeState:"clean", BuildDate:"2016-11-12T05:16:27Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by jessfraz** [comment]
have we gotten to the bottom of why this is?

**Comment by jessfraz** [comment]
well that answered my question lol

**Pull request merged**